Update rmi logic for canonical references

dmcgowan opened the pull request:
Updates the rmi code to treat canonical references as related to tagged references from the same repository during deletion. Canonical references with a different repository name will be treated as separate references. Updates the remove by ID logic to still remove an image if there is a single tag reference and only canonical references to the same repository remaining.

This was discussed as part of #23503. The `docker rmi` logic needs to be smarter about cleaning up canonical references before canonical references can always added on pull and push.

Ping @dnephin @aaronlehmann

dmcgowan committed:
Update rmi logic for canonical references

Updates the rmi code to treat canonical references as related to tagged references from the same repository during deletion.
Canonical references with a different repository name will be treated as separate references.
Updates the remove by ID logic to still remove an image if there is a single tag reference and only canonical references to the same repository remaining.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

aaronlehmann commented:
I think this demands an integration test.

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -105,26 +105,34 @@ func (daemon *Daemon) ImageDelete(imageRef string, force, prune bool) ([]types.I
-		// references so that they don't prevent removal of the image.
+		// to the same repository are digest references, delete the
+		// remaining repository references so that they don't prevent
+		// removal of the image.
```
Let's clarify the "so they don't prevent removal of the image" part.

As I understand it, there may well be other references to the image, so we aren't necessarily deleting it. In many cases, we'll be removing the digest for cleanliness.

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -135,11 +143,11 @@ func (daemon *Daemon) ImageDelete(imageRef string, force, prune bool) ([]types.I
+		// If an ID reference was given AND there is at most one
+		// repository tag reference to the image AND references within
+		// exactly one repository then we will want to remove all
+		// remaining references.
```
Can you clarify "references within exactly one repository"?

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -148,21 +156,55 @@ func (daemon *Daemon) ImageDelete(imageRef string, force, prune bool) ([]types.I
+// isSingleReference tests a list of references to see if either
+// references from multiple repositories are represented or if
+// there is only a single tag reference. Canonical references to
+// the same repository as a tag are ignored.
```
What does the function return?

aaronlehmann commented:
This looks correct to me. I found it quite hard to review, though.

I would really like to see one or more integration tests.

aaronlehmann commented:
Added to 1.12.0 milestone since this is blocking #23503.

dnephin commented on daemon/image_delete.go:
```diff
@@ -148,21 +156,55 @@ func (daemon *Daemon) ImageDelete(imageRef string, force, prune bool) ([]types.I
+		if canonical.Name() != singleRef.Name() {
+			return false
+		}
+	}
```
Maybe store `canonicalRefs.Name()` in a `map[string]bool` so this could be just 

```
_, ok := canonicalRefs[singleRef.Name()]
return ok
```

?

dmcgowan replied:
`map[string]bool` is a good idea, at the end I can just test whether there is more than one element in it though and bail earlier if there is multiple tags.

dmcgowan replied:
I think that is better, I will also need to test the length, `return ok && len(canonicalRefs) == 1` since this is testing there is not entries from other repositories as well.  Going to update the comment to make that clearer

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "image-delete-last-repo-ref" git@github.com:dmcgowan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "image-delete-last-repo-ref" git@github.com:dmcgowan/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

dmcgowan committed:
Add tests for rmi

Add integration test for removing by image id with tag and digest reference to the same repository.
Add integration test to ensure only tag to other repository remains after deleting tag with accompanying digest reference.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
@aaronlehmann added a test and attempted to clarify comments.

@dnephin simplified the `isSingleReference` logic using your suggestion.

tonistiigi commented:
SGTM

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -148,21 +154,48 @@ func (daemon *Daemon) ImageDelete(imageRef string, force, prune bool) ([]types.I
+		return len(repoRefs) == 1
+	}
+	var singleRef reference.Named
+	canonicalRefs := map[string]bool{}
```
Since this is being used as a set, the type should be `map[string]struct{}`.

aaronlehmann commented on integration-cli/docker_cli_by_digest_test.go:
```diff
@@ -412,6 +417,37 @@ func (s *DockerRegistrySuite) TestDeleteImageWithDigestAndTag(c *check.C) {
+	// rmi should have deleted repoTag and image reference, but left repoTag2
+	inspectField(c, repoTag2, "Id")
+	_, err = inspectFieldWithError(imageReference, "Id")
+	c.Assert(err, checker.NotNil, check.Commentf("image reference should have been deleted"))
```
Confirm that `repoTag` is removed as expected?

aaronlehmann commented:
LGTM

LK4D4 commented:
LGTM
